Simperium/SPCoreDataStorage.m

```diff
@@ -485,14 +492,45 @@ - (void)mainContextObjectsDidChange:(NSNotification *)notification {
 
 - (void)addObserversForMainContext:(NSManagedObjectContext *)moc {
     NSNotificationCenter* nc = [NSNotificationCenter defaultCenter];
-    [nc addObserver:self selector:@selector(managedContextWillSave:) name:NSManagedObjectContextWillSaveNotification object:moc];
-    [nc addObserver:self selector:@selector(mainContextDidSave:) name:NSManagedObjectContextDidSaveNotification object:moc];
-    [nc addObserver:self selector:@selector(mainContextObjectsDidChange:) name:NSManagedObjectContextObjectsDidChangeNotification object:moc];
+    [nc addObserver:self selector:@selector(mainContextWillSave:)           name:NSManagedObjectContextWillSaveNotification         object:moc];
+    [nc addObserver:self selector:@selector(mainContextDidSave:)            name:NSManagedObjectContextDidSaveNotification          object:moc];
+    [nc addObserver:self selector:@selector(mainContextObjectsDidChange:)   name:NSManagedObjectContextObjectsDidChangeNotification object:moc];
 }
 
 
 #pragma mark - Children MOC Notification Handlers
 
+- (void)childrenContextWillSave:(NSNotification*)notification {
+
+    // Obtain Permanent ID's!
+    NSManagedObjectContext *childrenContext = (NSManagedObjectContext *)notification.object;
+    [self obtainPermanentIDsForInsertedObjectsInContext:childrenContext];
+    
+    // Get the deleted ManagedObject ID's
+    NSMutableSet *workerDeletedIds = [NSMutableSet set];
+    for (NSManagedObject *object in childrenContext.deletedObjects) {
+        [workerDeletedIds addObject:object.objectID];
+    }
+    
+    if (workerDeletedIds.count == 0) {
+        return;
+    }
+    
+    // NOTE:
+    // Deleting an entity in a Children MOC, while there was a faulted reference in the MainMOC, might trigger a NSObjectInaccessibleException (#436).
+    // Workaround: we'll make sure the objects are actually loaded into the WriterMOC. This will effectively prevent an exception,
+    // and the object will get removed as soon as the DidSave note is merged.
+
+    [self.writerManagedObjectContext performBlockAndWait:^{
+        for (NSManagedObjectID *objectID in workerDeletedIds) {
+            NSManagedObject *writerMO = [self.writerManagedObjectContext existingObjectWithID:objectID error:nil];
+            if (writerMO.isFault) {
+                [writerMO willAccessValueForKey:nil];
```

Comment:
This does seem dirty but if it works its better than a crash. :smile: I can't see this being a problem if Apple ever fixes it either.